src/Orleans.Core/Providers/LegacyProviderConfigurator.cs

```diff
@@ -0,0 +1,168 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Diagnostics;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Orleans.LogConsistency;
+using Orleans.Runtime.Configuration;
+using Orleans.Storage;
+using Orleans.Streams;
+using System.Collections.Generic;
+using System.Linq;
+using Orleans.Runtime;
+
+namespace Orleans.Providers
+{
+    public class LegacyProviderConfigurator
+    {
+        public const string InitStageName = "ProviderInitStage";
+        public const string StartStageName = "ProviderStartStage";
+        // Optional task scheduling behavior, may not always be set.
+        internal delegate Task ScheduleTask(Func<Task> taskFunc);
+    }
+
+    internal static class LegacyProviderConfigurator<TLifeCycle>
+        where TLifeCycle : ILifecycleObservable
+    {
+        /// <summary>
+        /// Legacy way to configure providers. Will need to move to a legacy package in the future
+        /// </summary>
+        /// <returns></returns>
+        internal static void ConfigureServices(IDictionary<string, ProviderCategoryConfiguration> providerConfigurations, IServiceCollection services, int defaultInitStage, int defaultStartStage)
+        {
+            // if already added providers or nothing to add, skipp
+            if (services.Any(s => s.ServiceType == typeof(ProviderTypeLookup))
+                || providerConfigurations.Count == 0)
+                return;
+
+            services.AddSingleton<ProviderTypeLookup>();
+
+            foreach (var providerGroup in providerConfigurations.GroupBy(p => p.Key))
+            {
+                if (providerGroup.Key == ProviderCategoryConfiguration.STREAM_PROVIDER_CATEGORY_NAME)
+                {
+                    foreach (IProviderConfiguration providerConfig in providerGroup.SelectMany(kvp => kvp.Value.Providers.Values))
+                    {
+                        RegisterProvider<IStreamProvider>(providerConfig, services, defaultInitStage, defaultStartStage);
+                    }
+                }
+                else if (providerGroup.Key == ProviderCategoryConfiguration.STORAGE_PROVIDER_CATEGORY_NAME)
+                {
+                    services.AddSingleton<IStorageProvider>(sp => sp.GetServiceByName<IStorageProvider>(ProviderConstants.DEFAULT_STORAGE_PROVIDER_NAME));
```

Comment:
Can provider groups be specified multiple times? I'm not sure, but we should add this exactly once regardless of whether or not a storage provider with `DEFAULT_STORAGE_PROVIDER_NAME` has been registered.

i.e, I believe we should move it out of the loop.

Reply:
We perform a groupby, so this should only happen once.